controllers/topology/internal/templateutil/template.go

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templateutil
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CalculateTemplatesInUse returns all templates referenced in non-deleting MachineDeployment and MachineSets.
+func CalculateTemplatesInUse(md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (map[string]bool, error) {
+	templatesInUse := map[string]bool{}
+
+	// Templates of the MachineSet are still in use if the MachineSet is not in deleting.
+	for _, ms := range msList {
+		if !ms.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		bootstrapRef := ms.Spec.Template.Spec.Bootstrap.ConfigRef
+		infrastructureRef := &ms.Spec.Template.Spec.InfrastructureRef
+		if err := addTemplateRef(templatesInUse, bootstrapRef, infrastructureRef); err != nil {
+			// TODO(sbueringer) use KRef after 5153 has been merged
+			return nil, errors.Wrapf(err, "failed to add templates of %s/%s to templatesInUse", ms.Kind, ms.Name)
```

Comment:
note: I'm seen this `%s/%s, ms.Kind, ms.Name` pattern  all around in the recent topology PRs, we should probably have a helper.

Reply:
Yeah we should, but I have to get #5153 merged first :/ I assume it will happen before this PR so I can replace the TODOs

Reply:
nvm I see that's in 5153 https://github.com/kubernetes-sigs/cluster-api/pull/5153/files#diff-7537659f3b161f9836cb9e15e042a3c3ea4526e772450770ce9eda4b665f0988R123